File: go/leetcode/24. Swap Nodes in Pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    if head == nil {return nil}
    if head.Next == nil {return head}
    
    cur := head.Next
    head.Next = cur.Next
    cur.Next = head
    head = cur
    
    head.Next.Next = swapPairs(head.Next.Next)
    return head
}

File: c++/leetcode/24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  ListNode *swapPairs(ListNode *head)
  {
    return swapp(head);
  };
  ListNode *swapp(ListNode *);
};

ListNode *Solution::swapp(ListNode *head)
{
  if (!head)
    return nullptr;
  if (!head->next)
    return head;

  ListNode *cur = head->next;
  head->next = cur->next;
  cur->next = head;
  head = cur;
  head->next->next = swapp(head->next->next);

  return head;
}